### PR TITLE
added version protocol

### PR DESCRIPTION
**Pull request by PATAPON-NULL**
## 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have read and followed the `CONTRIBUTING.md` file
- [ ] I have applied the .clang-format file to my changes
- [ ] I have commented my code using Doxygen style comments (see Contributing MD for examples), particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the design documentation 
- [ ] I have updated any relevant parameters in the internal ICD
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If this is a bug-fix, I have added tests that can catch similar bugs in the future
- [ ] New and existing unit tests pass locally with my changes
- [ ] For cfe apps I have reserved an empty address for my msgs_ids and perf_ids from the [tracking list](https://docs.google.com/spreadsheets/d/1yRUhNTRBOiGLswlsWftNaYlwGKDeNQ_fCnDYFrc_7fU/edit#gid=772812381) and checked that it does not clash with other apps
- If any message is changed, removed, or added:
  - [ ] Please contact `#software` on Slack about the change(s)
  - [ ] **ONLY IF TELEMETRY MESSAGE** (for now): `./message-utilities/src/utilities.c` is updated with the correct field(s) properly marshalled to JSON **(IMPORTANT)**
  - If a new struct/message is added:
    - [ ] There is a timestamp included via either an explicit timestamp variable in the struct or buried in the a tlmHeader
  - If any of the existing struct's properties change in any way (property name change, property moves around in the struct, property removed):
    - [ ] **Breaking change** is checkmarked above as any change will affect every software downstream

### CODE REVIEW

File: fsw/public_inc/master_comms_bus_protocol.h [hunk 1]
```diff
@@ -94,4 +96,43 @@ static_assert(((MSP_HEALTH_PAYLOAD_LEN % 2) == 0),
...
+    MAX_GITTAG_LEN = 40,
+    MAX_TIME_LEN = 25,
+} version_char_len_t;
+
+typedef struct {
+    char git_hash[GITHASH_LEN];        // git commit hash
+    char git_tag[MAX_GITTAG_LEN];      // git tag information NO_TAG if there's not tag
+                                       // associate wit the commit
+    char built_time[MAX_TIME_LEN];     // binary built time
+} msp_version_payload_t;
```
Comment (tenzincrouch) [inline fsw/public_inc/master_comms_bus_protocol.h:111]: @PATAPON-NULL - lets add the additional fields we discussed even if they will be temporarily left blank

File: fsw/public_inc/master_comms_bus_protocol.h [hunk 2]
```diff
@@ -94,4 +96,43 @@ static_assert(((MSP_HEALTH_PAYLOAD_LEN % 2) == 0),
               "msp_health_payload_t struct not 16 bit aligned");
 #endif
 
+/* version messages */
+typedef enum {
+    GITHASH_LEN =  9,
+    MAX_GITTAG_LEN = 40,
+    MAX_TIME_LEN = 25,
+} version_char_len_t;
```
Comment (tenzincrouch) [inline fsw/public_inc/master_comms_bus_protocol.h:104]: Maybe make it clear these are string length?

File: fsw/public_inc/master_comms_bus_protocol.h [hunk 3]
```diff
@@ -94,4 +96,43 @@ static_assert(((MSP_HEALTH_PAYLOAD_LEN % 2) == 0),
...
+
+typedef struct {
+    char git_hash[GITHASH_LEN];        // git commit hash
+    char git_tag[MAX_GITTAG_LEN];      // git tag information NO_TAG if there's not tag
+                                       // associate wit the commit
+    char built_time[MAX_TIME_LEN];     // binary built time
```
Comment (tenzincrouch) [inline fsw/public_inc/master_comms_bus_protocol.h:110]: probably worth making clear that these are ascii strings

File: fsw/public_inc/master_comms_bus_protocol.h [hunk 4]
```diff
@@ -94,4 +96,43 @@ static_assert(((MSP_HEALTH_PAYLOAD_LEN % 2) == 0),
...
+    uint16_t checksum;
+    uint16_t __padding;   // for consistent footer on MSPs commands
+} msp_version_cmd_t;
+#if (__STDC_VERSION__ >= 201112L)   // check if compiling with C11 or newer
+// Preprocessor struct size and alignment checks
+static_assert((12 == MSP_HEALTH_PAYLOAD_LEN),
+              "msp_health_payload_t struct size incorrect (expected 4 bytes)");
+
+static_assert(((MSP_HEALTH_PAYLOAD_LEN % 4) == 0),
+              "msp_health_payload_t struct not 32 bit aligned");
```
Comment (tenzincrouch) [inline fsw/public_inc/master_comms_bus_protocol.h:135]: I think you will need to do this type of message struct for the version telemetry as well. 

For example see the power switching messages - https://github.com/PlanetaryRobotics/moonranger-rover-messages/blob/main/fsw/public_inc/power_switching_msgs.h

the payloads og the telemetry messages and commands are both put in _tlm_t or _cmd_t structs from line 131 onwards.

**Review by tenzincrouch — COMMENTED** [review]
Great start. I added a few comments to address.

**Review by tenzincrouch — APPROVED** [review]
LGTM

**Comment by epicstar** [comment]
FYI @tenzincrouch @PATAPON-NULL this actually breaks MCS downstream. Can you please aid how we can attempt to fix this?

https://github.com/PlanetaryRobotics/moonranger-mission-control/actions/runs/3535451016/jobs/5933489905

(Note @rohitp27 looks like dependabot is working well)

**Comment by PATAPON-NULL** [comment]
> FYI @tenzincrouch @PATAPON-NULL this actually breaks MCS downstream. Can you please aid how we can attempt to fix this?
> 
> https://github.com/PlanetaryRobotics/moonranger-mission-control/actions/runs/3535451016/jobs/5933489905
> 
> (Note @rohitp27 looks like dependabot is working well)
@epicstar apologize for my indiscretion [https://github.com/PlanetaryRobotics/moonranger-rover-messages/commit/8a2b99244d94340cdcabbca07f54ee6f2fb2ad5c](url) please refer the changes in this commit and see it will help

**Comment by epicstar** [comment]
@PATAPON-NULL your fix on `scratch/MSP_version` fixes us. Thank you!!!! Please create a PR and I'll approve.

**Pull request merged**